src/ipa/rpi/controller/rpi/awb_nn.cpp

```diff
@@ -0,0 +1,505 @@
+			b_diff[idx] = pixel.B - pixel.G;
+		}
+	}
+
+	std::vector<double> min_r_diff(total_pixels);
+	std::vector<double> min_b_diff(total_pixels);
+
+	for (int i = 0; i < 32; i++) {
+		for (int j = 0; j < 32; j++) {
+			int idx = i * 32 + j;
+
+			double min_abs_r = std::numeric_limits<double>::max();
+			double min_abs_b = std::numeric_limits<double>::max();
+
+			double selected_r = 0.0;
+			double selected_b = 0.0;
+
+			for (int di = -1; di <= 1; di++) {
+				for (int dj = -1; dj <= 1; dj++) {
+					double r = getElementPadded(r_diff, i + di, j + dj);
+					double b = getElementPadded(b_diff, i + di, j + dj);
+
+					if (std::abs(r) < min_abs_r) {
+						min_abs_r = std::abs(r);
+						selected_r = r;
+					}
+
+					if (std::abs(b) < min_abs_b) {
+						min_abs_b = std::abs(b);
+						selected_b = b;
+					}
+				}
+			}
+			min_r_diff[idx] = selected_r;
+			min_b_diff[idx] = selected_b;
+		}
+	}
+
+	for (int i = 0; i < 32; i++) {
+		for (int j = 0; j < 32; j++) {
+			int idx = i * 32 + j;
+
+			const AwbNN::RGB &original_pixel = rgb[idx];
+			double g = original_pixel.G;
+			AwbNN::RGB new_pixel(g + min_r_diff[idx], g, g + min_b_diff[idx]);
+			new_rgb[idx] = new_pixel;
+		}
+	}
+
+	return new_rgb;
+}
+
+AwbNN::RGB AwbNN::processZone(AwbNN::RGB zone, float lux, float red_gain, float blue_gain)
+{
+	/*
+	 * Renders the pixel at 5000K temperature then takes the log of the
+	 * r/g and b/g ratios. The log is then scaled to 0-1 and clamped.
+	 * These are then combined with the lux to get the input for the.
+	 * neural network.
+	 */
+	RGB zone_gains = zone;
+
+	zone_gains.R *= red_gain;
+	zone_gains.G *= 1.0;
+	zone_gains.B *= blue_gain;
+
+	RGB zone_ccm = zone_gains;
+
+	zone_ccm.R = nnConfig_.ccm[0] * zone_gains.R + nnConfig_.ccm[1] * zone_gains.G + nnConfig_.ccm[2] * zone_gains.B;
+	zone_ccm.G = nnConfig_.ccm[3] * zone_gains.R + nnConfig_.ccm[4] * zone_gains.G + nnConfig_.ccm[5] * zone_gains.B;
+	zone_ccm.B = nnConfig_.ccm[6] * zone_gains.R + nnConfig_.ccm[7] * zone_gains.G + nnConfig_.ccm[8] * zone_gains.B;
+
+	RGB zone_log;
+
+	double green = std::clamp(zone_ccm.G, 0.1, 1.0);
+
+	double r_ratio = zone_ccm.R / green;
+	double b_ratio = zone_ccm.B / green;
+
+	r_ratio = std::clamp(r_ratio, 0.01, 1.0);
+	b_ratio = std::clamp(b_ratio, 0.01, 1.0);
+
+	zone_log.R = std::log(r_ratio);
+	zone_log.G = std::log(lux) / 10;
+	zone_log.B = std::log(b_ratio);
+
+	zone_log.R += 1.0;
+	zone_log.B += 1.0;
+
+	zone_log.R /= 2.0;
+	zone_log.B /= 2.0;
+
+	zone_log.R = std::clamp(zone_log.R, 0.0, 1.0);
+	zone_log.G = std::clamp(zone_log.G, 0.0, 1.0);
+	zone_log.B = std::clamp(zone_log.B, 0.0, 1.0);
+
+	return zone_log;
+}
+
+static float interpolateOutput(float *output_data, int num_outputs, float min_temp, float max_temp)
+{
+	float total_value = 0.0;
+	float total_weight = 0.0;
+
+	for (int i = 0; i < num_outputs; i++) {
+		float temp = min_temp + i * (max_temp - min_temp) / (num_outputs - 1);
+		float weight = output_data[i];
+		weight *= weight;
```

Comment:
Interesting. Makes all the weights +ve, among other things! I wonder a bit about putting a softmax on the end of the network instead, is that something we've tried?